spec.html

```diff
@@ -11059,6 +11060,44 @@ <h1>Examples of Automatic Semicolon Insertion</h1>
       <p>In the circumstance that an assignment statement must begin with a left parenthesis, it is a good idea for the programmer to provide an explicit semicolon at the end of the preceding statement rather than to rely on automatic semicolon insertion.</p>
     </emu-clause>
   </emu-clause>
+  <emu-clause id="sec-hazards-of-automatic-semicolon-insertion">
+    <h1>Hazards of Automatic Semicolon Insertion</h1>
+    <em>This section is non-normative.</em>
+    <p>ECMAScript programs can be written in a style with very few semicolons by relying on automatic semicolon insertion. As described above, semicolons are not inserted at every newline, and automatic semicolon insertion can depend on multiple tokens across line terminators.</p>
+
+    <p>As new syntactic features are added to ECMAScript, additional grammar productions could be added that cause lines relying on automatic semicolon insertion preceding them to change grammar productions when parsed. In order to prevent cases where automatic semicolon insertion could change by adding source text outside of a grammar productions existing text boundaries, explicit semicolon use is recommended.</p>
```

Comment:
> that cause lines relying on automatic semicolon insertion preceding them to change grammar productions when parsed ... In order to prevent cases where automatic semicolon insertion could change by adding source text outside of a grammar productions existing text boundaries

I really dislike this phrasing. From my understanding (stealing @bmeck's excellent example):

```js
class A {
  [Symbol.iterator]() {}
}

class A {
  B // Relies on ASI, causing Syntax Error
  [Symbol.iterator]() {} // Must now be rewritten to `;[Symbol.iterator]`
}
```

It seems more clear to me if we stated it in terms of the newly added line omitting the semi (making it rely on ASI) affecting the already existing-and-valid line:

> ... additional grammar productions could be added that cause the next line to change grammar productions when parsed.

Reply:
The recommendation for me comes from a different sort of risk assessment, but we can rephrase it somewhat.

The crux of the recommendation is that given 2 grammar productions that are subject to ASI (lets use Statements for now since class fields are not here yet) `a` and `b`. `a` + `b` = `c` can alter the combination from 2 statements to a single statement `c` with the grammar production changing.

A more concrete example:

```sh
echo 'let x = 1\n' > a
echo '(function () {})()\n' > b
cat a b > c
```

There are 2 places of risk assessment.

1. preventing `a` from being combined with whatever `b` is. 
2. preventing `b` from being combined with whatever `a` is.

Both of these places of risk are able to be acted upon in a preventative manner by putting a `;` in the proper locations. I would like any rephrasing to be able to convey both places of risk. I have no suggestions off the top of my head though.

Reply:
s/productions/production's/

Editorial suggestion: This is getting a bit wordy. OTOH it'd be nice to have something here that's widely quotable. What if we switch this paragraph and then following one, and refer to ASI hazards in this one?

Reply:
This is what StandardJS already recommends - putting semicolon in required locations (before `(` and `[` if it's at the beginning of line), so to prevent this issue. You autoformat your code:

```bash
echo '(function () {})()\n ' | prettier-standard
;(function () {})()
```

Auto-formatting also deals with the class example by leaving semicolon:

```bash
echo 'class A {
  B;
  [Symbol.iterator]() {}
}' | prettier-standard
class A {
  B;
  [Symbol.iterator] () {}
}
```


Instead enforcing putting semicolons everywhere, enforce putting them only in such locations.

Reply:
@sheerun main argument of StandardJS is that you do not need to remember all those semicolon locations, you just use the style, it's easy, etc.. But from what you're saying -- you need to know those locations and remember to insert semicolons. So not sure how this is related to that particular style or what it recommends. For me, main point here should be -- you cannot write semicolon-free JavaScript, whatever style you use.

Reply:
> you cannot write semicolon-free JavaScript

However, you can write *mostly* semicolon-free JS, with the remaining semicolons often pointing toward awkward constructs.

Reply:
@sheerun and this PR has a recommendation of the same (plus the other locations left out). The PR is very specific about where it is recommending things.

@j-f1 @NekR this PR is not about style nor about ease of learning. It provides documentation and an actionable response to specific situations.